Support for `polars::prelude::LiteralValue::Decimal(i128)` literal type

### Description

We can construct an Int128Chunked from an array of i128 values, which can be sequentially transformed into a series of Decimals with a default precision of 38 and a default scale of 0. See the following piece of code:

# Code
```rust
use polars::prelude::*;

fn main() {
    let a = Int128Chunked::from_vec(
        "", 
        vec![-2_i128, 2, -6, 10, 2, -9]
    );

    let lazy_data = df!(
        "a" => a,
        "b" => [20, 10, 30, 40, 50, 60]
    ).unwrap().lazy();

    let data = lazy_data.select(
        &[(lit(2) * col("a")).alias("a2"), (lit(-3) * col("b")).alias("-b3")]
    ).collect().unwrap();

    println!("Data: {:?}", data);
}
```

```
[dependencies]
polars = { version = "0.32.1", features = ["lazy", "bigidx", "dtype-decimal"]}
```

# Output for the previous code
```
Data: shape: (6, 2)
┌───────────────┬──────┐
│ a2            ┆ -b3  │
│ ---           ┆ ---  │
│ decimal[38,0] ┆ i32  │
╞═══════════════╪══════╡
│ -4            ┆ -60  │
│ 4             ┆ -30  │
│ -12           ┆ -90  │
│ 20            ┆ -120 │
│ 4             ┆ -150 │
│ -18           ┆ -180 │
└───────────────┴──────┘

```

However, we cannot construct a literal from an i128 value with a default precision of 38 and a default scale of 0 since there is no enumerated type `polars::prelude::LiteralValue::Decimal(i128)` available for this purpose. For instance, using a `lit(-3_i128)` instead of `lit(3)`, we get the following error:

```rust
   Compiling polars-test v0.1.0 (/home/joe/Documents/polars-test)
error[E0277]: the trait bound `i128: polars::prelude::Literal` is not satisfied
   --> src/main.rs:15:49
    |
15  |         &[(lit(2) * col("a")).alias("a2"), (lit(-3_i128) * col("b")).alias("-b3")]
    |                                             --- ^^^^^^^ the trait `polars::prelude::Literal` is not implemented for `i128`
    |                                             |
    |                                             required by a bound introduced by this call
    |
    = help: the following other types implement trait `polars::prelude::Literal`:
              i32
              i64
              u32
              u64
              f32
              f64
note: required by a bound in `polars::prelude::lit`
   --> /home/joe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/polars-plan-0.32.1/src/logical_plan/lit.rs:309:15
    |
309 | pub fn lit<L: Literal>(t: L) -> Expr {
    |               ^^^^^^^ required by this bound in `lit`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `polars-test` (bin "polars-test") due to previous error
```

As a solution, we can use `polars::prelude::LiteralValue::Series(SpecialEq::<Series>::new(s))`. However, it would be nice to have the `polars::prelude::LiteralValue::Decimal(i128)` enum, given that other data types already have their own enum types.

--- a/py-polars/tests/unit/datatypes/test_decimal.py
+++ b/py-polars/tests/unit/datatypes/test_decimal.py
@@ -659,3 +659,22 @@ def test_unique() -> None:
     assert ser.unique().to_list() == [D("1.1"), D("2.2")]
     assert ser.n_unique() == 2
     assert ser.arg_unique().to_list() == [0, 2]
+
+
+def test_groupby_agg_single_element_11232() -> None:
+    data = {"g": [-1], "decimal": [-1]}
+    schema = {"g": pl.Int64, "decimal": pl.Decimal(38, 0)}
+    result = (
+        pl.LazyFrame(data, schema=schema)
+        .group_by("g", maintain_order=True)
+        .agg(pl.col("decimal").min())
+        .collect()
+    )
+    expected = pl.DataFrame(data, schema=schema)
+    assert_frame_equal(result, expected)
+
+
+def test_decimal_from_large_ints_9084() -> None:
+    numbers = [2963091539321097135000000000, 25658709114149718824803874]
+    s = pl.Series(numbers, dtype=pl.Decimal)
+    assert s.to_list() == [D(n) for n in numbers]

--- a/py-polars/tests/unit/datatypes/test_decimal.py
+++ b/py-polars/tests/unit/datatypes/test_decimal.py
@@ -663,7 +663,7 @@ def test_unique() -> None:
 
 def test_groupby_agg_single_element_11232() -> None:
     data = {"g": [-1], "decimal": [-1]}
-    schema = {"g": pl.Int64, "decimal": pl.Decimal(38, 0)}
+    schema = {"g": pl.Int64(), "decimal": pl.Decimal(38, 0)}
     result = (
         pl.LazyFrame(data, schema=schema)
         .group_by("g", maintain_order=True)

--- a/py-polars/tests/unit/functions/as_datatype/test_concat_list.py
+++ b/py-polars/tests/unit/functions/as_datatype/test_concat_list.py
@@ -172,3 +172,23 @@ def test_concat_list_empty() -> None:
 def test_concat_list_empty_struct() -> None:
     df = pl.DataFrame({"a": []}, schema={"a": pl.Struct({"b": pl.Boolean})})
     df.select(pl.concat_list("a"))
+
+
+def test_cross_join_concat_list_18587() -> None:
+    lf = pl.LazyFrame({"u32": [0, 1], "str": ["a", "b"]})
+
+    lf1 = lf.select(pl.struct(pl.all()).alias("1"))
+    lf2 = lf.select(pl.struct(pl.all()).alias("2"))
+    lf3 = lf.select(pl.struct(pl.all()).alias("3"))
+
+    result = (
+        lf1.join(lf2, how="cross")
+        .join(lf3, how="cross")
+        .select(pl.concat_list("1", "2", "3"))
+        .collect()
+    )
+
+    vals = [{"u32": 0, "str": "a"}, {"u32": 1, "str": "b"}]
+    expected = [[a, b, c] for a in vals for b in vals for c in vals]
+
+    assert result["1"].to_list() == expected

--- a/py-polars/tests/unit/functions/test_when_then.py
+++ b/py-polars/tests/unit/functions/test_when_then.py
@@ -736,3 +736,21 @@ def test_struct_when_then_broadcasting_combinations_19122(
             pl.when(pl.col.a.struct.field("x") != 0).then(rv).otherwise(None).alias("a")
         ),
     )
+
+
+def test_when_then_to_decimal_18375() -> None:
+    df = pl.DataFrame({"a": ["1.23", "4.56"]})
+
+    result = df.with_columns(
+        b=pl.when(False).then(None).otherwise(pl.col("a").str.to_decimal()),
+        c=pl.when(True).then(pl.col("a").str.to_decimal()),
+    )
+    expected = pl.DataFrame(
+        {
+            "a": ["1.23", "4.56"],
+            "b": ["1.23", "4.56"],
+            "c": ["1.23", "4.56"],
+        },
+        schema={"a": pl.String, "b": pl.Decimal, "c": pl.Decimal},
+    )
+    assert_frame_equal(result, expected)

--- a/py-polars/tests/unit/operations/namespaces/list/test_list.py
+++ b/py-polars/tests/unit/operations/namespaces/list/test_list.py
@@ -680,6 +680,12 @@ def test_list_to_struct() -> None:
     ).collect_schema() == {"n": pl.Unknown}
 
 
+def test_list_to_struct_all_null_12119() -> None:
+    s = pl.Series([None], dtype=pl.List(pl.Int64))
+    result = s.list.to_struct(fields=["a", "b", "c"]).to_list()
+    assert result == [{"a": None, "b": None, "c": None}]
+
+
 def test_select_from_list_to_struct_11143() -> None:
     ldf = pl.LazyFrame({"some_col": [[1.0, 2.0], [1.5, 3.0]]})
     ldf = ldf.select(

--- a/py-polars/tests/unit/operations/test_fill_null.py
+++ b/py-polars/tests/unit/operations/test_fill_null.py
@@ -1,3 +1,7 @@
+import datetime
+
+import pytest
+
 import polars as pl
 from polars.testing import assert_series_equal
 
@@ -68,3 +72,15 @@ def test_fill_null_decimal_with_int_14331() -> None:
     result = s.fill_null(0)
     expected = pl.Series("a", ["1.1", "0.0"], dtype=pl.Decimal(precision=None, scale=5))
     assert_series_equal(result, expected)
+
+
+def test_fill_null_date_with_int_11362() -> None:
+    match = "got invalid or ambiguous dtypes"
+
+    s = pl.Series([datetime.date(2000, 1, 1)])
+    with pytest.raises(pl.exceptions.InvalidOperationError, match=match):
+        s.fill_null(0)
+
+    s = pl.Series([None], dtype=pl.Date)
+    with pytest.raises(pl.exceptions.InvalidOperationError, match=match):
+        s.fill_null(1)

--- a/py-polars/tests/unit/operations/test_join_asof.py
+++ b/py-polars/tests/unit/operations/test_join_asof.py
@@ -1278,3 +1278,9 @@ def test_join_asof_no_exact_matches() -> None:
         "bid": [None, 51.97, None, None, None],
         "ask": [None, 51.98, None, None, None],
     }
+
+
+def test_join_asof_not_sorted_warning() -> None:
+    df = pl.DataFrame({"a": [1, 1, 1, 2, 2, 2], "b": [2, 1, 3, 1, 2, 3]})
+    with pytest.warns(UserWarning, match="asof join is not sorted"):
+        df.join_asof(df, on="b", by="a")

--- a/py-polars/tests/unit/series/test_equals.py
+++ b/py-polars/tests/unit/series/test_equals.py
@@ -297,3 +297,9 @@ def test_eq_missing_lists_arrays_19153(
         ),
         pl.Series([False, True, False]),
     )
+
+
+def test_equals_nested_null_categorical_14875() -> None:
+    dtype = pl.List(pl.Struct({"cat": pl.Categorical}))
+    s = pl.Series([[{"cat": None}]], dtype=dtype)
+    assert s.equals(s)

